[V3 Permissions] Find things uniquely for models

mikeshardmind opened the pull request:
### Type

- [ ] Bugfix
- [X] Enhancement
- [ ] New feature

### Description of the changes

This is a safety measure to prevent accidentally passing a user when you wanted a role with the same name. See discord discussion, and @Kowlin 

If it can't find it uniquely, it should tell them to use an id.

Tobotimus commented:
Needs a rebase before merging to be included in 3.0.0

Commit:
Let's put the changes on the correct branch this time

mikeshardmind commented:
:+1: Could have sworn I branched off of V3/release/3.0.0, but was branched off of FETCH_HEAD instead.

Commit:
`is_default` is a method, not a property

Tobotimus commented on redbot/cogs/permissions/converters.py:
```diff
@@ -6,6 +6,58 @@
 
 
+class GlobalUniqueObjectFinder(commands.Converter):
+    @staticmethod
```
Unfortunately static methods don't work as convert methods because they don't pass the `inspect.ismethod` predicate. They need to be class methods instead

Tobotimus commented on redbot/cogs/permissions/converters.py:
```diff
@@ -6,6 +6,58 @@
+            if len(maybe_matches) > 0:
+                raise commands.BadArgument(
+                    "Could not uniquely match that, please use an id or mention for this", arg
+                )
```
Message args to `commands.BadArgument` are user-facing strings, and thus need to be translated

Tobotimus reviewed (dismissed):
This is a good change but could be done more efficiently. If the user passes an ID or mention it should find the object in constant time, or O(guild count) time for global.
This also won't find members from their nickname, if supplied.

Commit:
performance boost this

Commit:
fix the other stuff

Commit:
it's possible to mention things the bot can't see

Tobotimus committed:
Improvements and fixes

- Mention regex fixed (was previously matching mentions but not capturing ID)
- Fixed AttributeError being raised when finding an object by name
- Speed optimisations (models found in O(1) operations when ID or mention, except for roles in global search which is O(n))
- Memory optimisations (less copying and creation of collections, use of itertools and references to existing sequences)
- Clearer user feedback for bad arguments

Signed-off-by: Toby Harradine <tobyharradine@gmail.com>

Commit:
Merge pull request #4 from Tobotimus/V3/find_objects_uniquely

Improvements and fixes

Commit:
a few more tweaks for performance
